rfcs/0001-masonry-backend.md

```diff
@@ -0,0 +1,166 @@
+The `View` trait has an `Element` associated type, which depends on the platform layer, and represents a persisten UI entity. Every time app data changes, a new View is produced. That View is diffed against the app's previous View, and all changes are applied to matching elements: if the new view has a child view that wasn't in the previous version, a new element is added; if the new view no longer has a child view that was in the previous version, the matching element is removed, and so on.
+
+(In web frameworks that use a virtual DOM, this process is sometimes known as "reconciliation".)
+
+Xilem has multiple platform layers. `xilem_web` is based on the web DOM, whereas `xilem_native` is based on the Masonry crate, which provides a native widget tree and some Developer Experience features.
+
+Note that, while the platform layer is sometimes referred to as a "backend", it's somewhat tightly coupled with the reactive layer.
+
+`xilem_web`, `xilem_native` and any other platforms will each have their own version of the `View` trait, the `ViewSequence` trait, the `AnyView` trait, the `BoxedView` type, and so on.
+
+### `xilem_native`
+
+The element type in `xilem_native` must implement `masonry::Widget`.
+
+Widget in Masonry is a trait that represent units of graphical composition, things like text and buttons and boxes and containers. To implement Widget, a type must:
+
+- Be `'static`.
+- Provide info about its children.
+- Handle user events.
+- Compute its layout.
+- Draw itself in a paint target.
+
+See the [Widget trait definition](https://github.com/PoignardAzur/masonry-rs/blob/3cfc2cddec05d48a31a9812bdda27792d336f03f/src/widget/widget.rs#L65-L199) for more details.
+
+
+### Directory structure
+
+The `https://github.com/linebender/xilem` repository has the following crates in its `crates/` directory:
+
+- `xilem_core/`
+- `xilem_web/`
+- `xilem_native/`
+- `masonry/`
+- `xilem/`, which would only be a placeholder crate pointing to `xilem_web/` or `xilem_native/`.
+
+
+## Implementation strategy
+
+The implementation will require several steps:
+
+- Refactoring Masonry to use the same dependencies Xilem currently uses, especially Glazier, Vello, and Parley insted of druid-shell and Piet.
+- Integrating AccessKit support.
+- Moving the Masonry codebase to the Xilem repository, ideally in a way that preserves commit history.
+- Porting the list of Github Issues in the masonry-rs repository to the Xilem repository.
+- Creating a new `xilem_native` crate on parity with the current Xilem crate.
+- Removing the current xilem crate.
+- Publishing both `xilem_native` and the placeholder `xilem` crate on crates.io.
+- Sharing ownership of the crates with Raph Levien.
+
+
+## Drawbacks
+
+### Layer separation
+
+This move decouples the reactive layer of xilem from its platform layer by putting them in separate crates.
+
+This means some features in `xilem_native` may be harder to implement because they won't have access to the backend's private code.
```

Comment:
I think this needs more detailed explanation, do you mean the whole runtime? Only the widgets? The xilem `Widget` trait?

I'm also not sure whether this should be delayed and just prototyped when deemed necessary (and based on that experience a new RFC or depending on the complexity just a PR opened), I don't see much benefit in splitting xilem in too many crates currently (mostly because of the orphan rule). AFAIU the widgets in masonry will be its own crate already anyway, so the likely biggest amount of code is already in a separate crate.

Reply:
Not sure if this is really resolved is it? (same with the other comments) I don't see where this is explained in more detail in the new version. I can only guess, that the temporary `xilem_masonry` will contain the runtime (i.e. is the temporary merger) and this will be the new `xilem` at some point

Reply:
I guess I could draw up a list of public items in `xilem` and split them between "frontend" and "backend". That would probably make the explanation a lot easier to grok.

Reply:
Yeah, figuring out the composition root might be a bit tricky, but I don't think it should be included in the RFC.

> You should probably include a lot what you've written here in the RFC itself

Will do. I do need to make these things more explicit.

> So that the dependency chain would look like this: runtime -> reactive/view -> widgets (masonry)

Most of the "runtime" as it currently exists will be defined in Masonry. The rough idea is that, when you create a Masonry window, you give it a bunch of callbacks (or a trait impl or whatever) to be called on every paint, input event, etc. So the composition root of Xilem will be in the callbacks passed to Masonry's own composition root.

Reply:
> Yeah, figuring out the composition root might be a bit tricky, but I don't think it should be included in the RFC.

I think this was the reason why things were unclear to me a little bit and thus the question (runtime vs reactive layer etc.), since it's interconnected above `widget/` and `view/` (mainly `App` and `AppTask`) and thus probably not cleanly separable between "platform" and "reactive" layer (or we have a different interpretation of "reactive" layer) with just 2 crates (`xilem` and `masonry`).

Sure a lot of the code in these modules can be laid out in masonry etc. but as you say details can be done in PRs (I think it will clear up when `xilem_masonry` is being implemented).

Reply:
I've added a new sub-section that mentions the source files and how they'll be affected, though what the composition root will look like is left a bit vague.

Let me know if you think it should be more clear/explicit!

Reply:
Yeah I think continuing discussion makes more sense in the relevant parts of the RFC